src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -328,15 +383,24 @@ public void CopyChangedValues(TStructuralType original)
                 }
                 else
                 {
-                    // Recursively patch the subtree.
-                    bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
-                    Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+                    //For Delta collection (Edmchangedobjectcoll), these will get called for each nested collection in delta                     
+                    if (deltaNestedResource is EdmChangedObjectCollection)
+                    {
+                        deltaNestedResource.CopyChangedValues(originalNestedResource);
+                    }
+                    else
+                    {
+                        //Recursively patch the subtree.
```

Comment:
Lines 386-397 mix logic for a contract.assert with production logic.  The production logic is always the same -- call CopyChangedValues.  All of the type checking should only be done in debug mode (i.e., should be part of the assert).  So you can rewrite 387-397 as:

Contract.Assert(deltaNestedResource is EdmChangedObjectCollection || TypedDelta.IsDeltaOfT(deltaNestedResource.GetType()), nestedResourceName + "should be EdmChangedObjectCollection, or Delta<T> with a corresponding type <T>, but is not.");
deltaNestedResource.CopyChangedValues(originalNestedResource);